File: trunk/app/src/com/kostmo/wallpaper/spiral/base/ArchimedeanSpiral.java
package com.kostmo.wallpaper.spiral.base;

import android.graphics.Path;
import android.graphics.Point;
import android.graphics.PointF;
import android.util.FloatMath;


public class ArchimedeanSpiral extends SpiralGenerator {

	// ========================================================================
	public ArchimedeanSpiral(int arcDivisions, int periods) {
		super(arcDivisions, periods);
		initialize();
	}

	// ========================================================================
	@Override
	float spiralLength(float theta) {
		float length = ( (theta * FloatMath.sqrt(1 + theta*theta)) + (float) Math.log(theta + FloatMath.sqrt(1 + theta*theta)) ) / 2;
		return length;
	}

	// ========================================================================
	public float getScale(Point screen_dimensions) {

		float half_diagonal = PointF.length(screen_dimensions.x, screen_dimensions.y)/2f;
		float max_radius = spiralRadius((this.periods - 1) * TWO_PI);

		return half_diagonal / max_radius;
	}
	
	
	// ========================================================================
	@Override
	float spiralSlope(float theta) {
		float radius = this.spiralRadius(theta);
		float cos = FloatMath.cos(theta);
		float sin = FloatMath.sin(theta);
		
		float dy = sin/TWO_PI + radius*cos;
		float dx = cos/TWO_PI - radius*sin;
		
		return dy/dx;
	}

	// ========================================================================
	@Override
	float spiralRadius(float theta) {
		return theta/TWO_PI;
	}

	// ========================================================================
	@Override
	public Path getPath() {
		return getSimplePath();
	}
}

File: trunk/parameter_exploration/archimedean_spiral.py
#! /usr/bin/env python

from math import pi, sin, cos
from base_spiral import BaseSpiral, run

# =============================================================================
class ArchimedeanSpiral(BaseSpiral):

	def __init__(self, segment_count, **kwargs):

		BaseSpiral.__init__(self, segment_count)

	# ---------------------------------------------------------------------
	def spiralLength(self, w):
		from math import log, sqrt
		return ( (w * sqrt(1 + w**2)) + log(w + sqrt(1 + w**2)) ) / 2

	# ---------------------------------------------------------------------
	def spiralSlope(self, w):
		dy = sin(w)/(2*pi) + self.spiralRadius(w)*cos(w)
		dx = cos(w)/(2*pi) - self.spiralRadius(w)*sin(w)
		return dy / dx

	# ---------------------------------------------------------------------
	def spiralRadius(self, theta):
		return theta/(2*pi)

# =============================================================================
if __name__ == "__main__":
	run(ArchimedeanSpiral)
